keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -649,89 +629,182 @@ endpoints:
+multi-cluster service to become accessible from within the supercluster. The
+domain name will be `<service>.<ns>.svc.supercluster.local`. 
+
+**SuperclusterIP services:** Requests to this domain name from within an importing
+cluster will resolve to the supercluster IP, which points to endpoints for pods
+within the underlying `Service`(s) across the supercluster.
+
+**Headless services:** Within an importing cluster, the supercluster domain name
+will have multiple `A`/`AAAA` records, each containing the address of a ready
+endpoint of the headless service. `<service>.<ns>.svc.supercluster.local` will
+resolve to the set of all ready pod IPs for the service.
+
+Pods backing a supercluster service may be addressed individually using the
+`<hostname>.<clustername>.<svc>.<ns>.svc.supercluster.local` format. Necessary
+records will be created based on each ready endpoint's hostname and the
+`multicluster.kubernetes.io/source-cluster` label on the `EndpointSlice`. This
+allows naming collisions to be avoided for headless services backed by identical
+`StatefulSets` deployed in multiple clusters.
+
+_Note: the total length of a FQDN is limited to 253 characters. Each label is
+independently limited to 63 characters, so users must choose host/cluster/service
+names to avoid hitting this upper bound._
+
+All service consumers must use the `*.svc.supercluster.local` name to enable
+supercluster routing, even if there is a matching `Service` with the same
+namespaced name in the local cluster. There will be no change to existing
+behavior of the `cluster.local` zone.
+
+#### EndpointSlice
+
+When a `ServiceExport` is created, this will cause `EndpointSlice` objects for
+the underlying `Service` to be created in each importing cluster within the supercluster,
+associated with the derived `ServiceImport`. One or more `EndpointSlice`
+resources will exist for the exported `Service`, with each `EndpointSlice`
+containing only endpoints from a single source cluster. These `EndpointSlice`
+objects will be marked as managed by the supercluster service controller, so
+that the endpoint slice controller doesn’t delete them. `EndpointSlices` will
+have an owner reference to their associated `ServiceImport`.
+
+```
+<<[UNRESOLVED]>>
+We have not yet sorted out scalability impact here. We hope the upper bound for
+imported endpoints + in-cluster endpoints will be ~= the upper bound for
+in-cluster endpoints today, but this remains to be determined.
+<<[/UNRESOLVED]>>
+```
+
+#### Endpoint TTL
+
+To prevent stale endpoints from persisting in the event that a cluster becomes
+unreachable to the supercluster controller, each `EndpointSlice` is associated
+with a lease representing connectivity with its source cluster. The supercluster
+service controller is responsible for periodically renewing the lease so long as
+the connection with the source cluster is confirmed alive. A separate
+controller, that may run inside each cluster, is responsible for watching each
```

Comment:
may run ?  Have we POC'ed this?

How will leases be named and discovered/linked to the ServiceImport?

Reply:
There is no POC yet, this is one of the things I'm hoping will solidify as implementations develop around the Alpha API. I've reworded this to be a recommendation, and suggested that the cluster-name and the `mulitcluster.k8s.io/source-cluster` label be used to find slices mapping to an expired cluster lease.